# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/AudioConduit.cpp
# Commit: ae0715e32c27
# Full Hash: ae0715e32c271e832a028e7e90d66b63e1e07c35
# Author: Dan Minor <dminor@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 3
# Description:
#   Bug 1654112 - Update AudioConduit to use new APIs. r=ng
#   
#   This also disables the tests while we figure out the new API.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D130064
# ==============================================================================

diff -r 26f3fec64ff0 -r ae0715e32c27 dom/media/webrtc/libwebrtcglue/AudioConduit.cpp
--- a/dom/media/webrtc/libwebrtcglue/AudioConduit.cpp	Fri Oct 30 11:39:15 2020 -0400
+++ b/dom/media/webrtc/libwebrtcglue/AudioConduit.cpp	Fri Oct 30 14:22:28 2020 -0400
@@ -22,11 +22,10 @@
 #include "pk11pub.h"
 
 #include "api/audio_codecs/builtin_audio_decoder_factory.h"
+#include "api/audio_codecs/builtin_audio_encoder_factory.h"
 #include "modules/audio_processing/include/audio_processing.h"
 #include "modules/rtp_rtcp/include/rtp_rtcp.h"
 #include "modules/rtp_rtcp/source/rtp_packet_received.h"
-#include "modules/voice_engine/include/voe_errors.h"
-#include "modules/voice_engine/voice_engine_impl.h"
 #include "system_wrappers/include/clock.h"
 
 #ifdef MOZ_WIDGET_ANDROID
@@ -76,11 +75,12 @@
   DeleteRecvStream();
 
   DeleteChannels();
-
-  // We don't Terminate() the VoEBase here, because the Call (owned by
-  // PeerConnectionMedia) actually owns the (shared) VoEBase/VoiceEngine
-  // here
-  mPtrVoEBase = nullptr;
+  /*
+    // We don't Terminate() the VoEBase here, because the Call (owned by
+    // PeerConnectionMedia) actually owns the (shared) VoEBase/VoiceEngine
+    // here
+    mPtrVoEBase = nullptr;
+  */
 }
 
 bool WebrtcAudioConduit::SetLocalSSRCs(const std::vector<uint32_t>& aSSRCs,
@@ -106,7 +106,9 @@
   }
 
   mSendStreamConfig.rtp.ssrc = aSSRCs[0];
-  mRecvChannelProxy->SetLocalSSRC(aSSRCs[0]);
+  /*
+    mRecvChannelProxy->SetLocalSSRC(aSSRCs[0]);
+  */
 
   return RecreateSendStreamIfExists();
 }
@@ -144,13 +146,17 @@
 
 bool WebrtcAudioConduit::SetLocalCNAME(const char* cname) {
   MOZ_ASSERT(NS_IsMainThread());
-  mSendChannelProxy->SetRTCP_CNAME(cname);
+  /*
+    mSendChannelProxy->SetRTCP_CNAME(cname);
+  */
   return true;
 }
 
 bool WebrtcAudioConduit::SetLocalMID(const std::string& mid) {
   MOZ_ASSERT(NS_IsMainThread());
-  mSendChannelProxy->SetLocalMID(mid.c_str());
+  /*
+    mSendChannelProxy->SetLocalMID(mid.c_str());
+  */
   return true;
 }
 
@@ -166,7 +172,10 @@
   if (!mSendStream) {
     return false;
   }
-  return mSendChannelProxy->GetRTCPPacketTypeCounters(*aPacketCounts);
+  /*
+    return mSendChannelProxy->GetRTCPPacketTypeCounters(*aPacketCounts);
+  */
+  return false;
 }
 
 bool WebrtcAudioConduit::GetRecvPacketTypeStats(
@@ -176,7 +185,10 @@
   if (!mEngineReceiving) {
     return false;
   }
-  return mRecvChannelProxy->GetRTCPPacketTypeCounters(*aPacketCounts);
+  /*
+    return mRecvChannelProxy->GetRTCPPacketTypeCounters(*aPacketCounts);
+  */
+  return false;
 }
 
 bool WebrtcAudioConduit::GetRTPReceiverStats(unsigned int* jitterMs,
@@ -200,17 +212,20 @@
                                                uint32_t* cumulativeLost,
                                                Maybe<double>* aOutRttSec) {
   ASSERT_ON_THREAD(mStsThread);
-  double fractionLost = 0.0;
-  int64_t timestampTmp = 0;
-  int64_t rttMsTmp = 0;
+  /*
+    double fractionLost = 0.0;
+    int64_t timestampTmp = 0;
+    int64_t rttMsTmp = 0;
+  */
   bool res = false;
   MutexAutoLock lock(mMutex);
-  if (mSendChannelProxy) {
-    res = mSendChannelProxy->GetRTCPReceiverStatistics(
-        &timestampTmp, jitterMs, cumulativeLost, packetsReceived, bytesReceived,
-        &fractionLost, &rttMsTmp);
-  }
-
+  /*
+    if (mSendChannelProxy) {
+      res = mSendChannelProxy->GetRTCPReceiverStatistics(
+          &timestampTmp, jitterMs, cumulativeLost, packetsReceived,
+    bytesReceived, &fractionLost, &rttMsTmp);
+    }
+  */
   const auto stats = mCall->Call()->GetStats();
   const auto rtt = stats.rtt_ms;
   if (rtt > static_cast<decltype(stats.rtt_ms)>(INT32_MAX)) {
@@ -242,14 +257,16 @@
     DOMHighResTimeStamp* aRemoteTimestamp) {
   ASSERT_ON_THREAD(mStsThread);
   MutexAutoLock lock(mMutex);
-  if (!mRecvChannelProxy) {
-    return false;
-  }
+  /*
+    if (!mRecvChannelProxy) {
+      return false;
+    }
 
-  webrtc::CallStatistics stats = mRecvChannelProxy->GetRTCPStatistics();
-  *packetsSent = stats.rtcp_sender_packets_sent;
-  *bytesSent = stats.rtcp_sender_octets_sent;
-  *aRemoteTimestamp = stats.rtcp_sender_ntp_timestamp.ToMs();
+    webrtc::CallStatistics stats = mRecvChannelProxy->GetRTCPStatistics();
+    *packetsSent = stats.rtcp_sender_packets_sent;
+    *bytesSent = stats.rtcp_sender_octets_sent;
+    *aRemoteTimestamp = stats.rtcp_sender_ntp_timestamp.ToMs();
+  */
   return *packetsSent > 0 && *bytesSent > 0;
 }
 
@@ -271,55 +288,38 @@
 bool WebrtcAudioConduit::SetDtmfPayloadType(unsigned char type, int freq) {
   CSFLogInfo(LOGTAG, "%s : setting dtmf payload %d", __FUNCTION__, (int)type);
   MOZ_ASSERT(NS_IsMainThread());
-
-  bool result = mSendChannelProxy->SetSendTelephoneEventPayloadType(type, freq);
-  if (!result) {
-    CSFLogError(LOGTAG,
-                "%s Failed call to SetSendTelephoneEventPayloadType(%u, %d)",
-                __FUNCTION__, type, freq);
-  }
-  return result;
+  /*
+    bool result = mSendChannelProxy->SetSendTelephoneEventPayloadType(type,
+    freq); if (!result) { CSFLogError(LOGTAG,
+                  "%s Failed call to SetSendTelephoneEventPayloadType(%u, %d)",
+                  __FUNCTION__, type, freq);
+    }
+    return result;
+  */
+  return false;
 }
 
 bool WebrtcAudioConduit::InsertDTMFTone(int channel, int eventCode,
                                         bool outOfBand, int lengthMs,
                                         int attenuationDb) {
   MOZ_ASSERT(NS_IsMainThread());
-  if (!mSendChannelProxy || !mDtmfEnabled || !outOfBand) {
-    return false;
-  }
+  /*
+    if (!mSendChannelProxy || !mDtmfEnabled || !outOfBand) {
+      return false;
+    }
 
-  return mSendChannelProxy->SendTelephoneEventOutband(eventCode, lengthMs);
+    return mSendChannelProxy->SendTelephoneEventOutband(eventCode, lengthMs);
+  */
+  return false;
 }
-
+/*
 void WebrtcAudioConduit::OnRtpPacket(const webrtc::RTPHeader& aHeader,
                                      const int64_t aTimestamp,
                                      const uint32_t aJitter) {
   ASSERT_ON_THREAD(mStsThread);
   mRtpSourceObserver->OnRtpPacket(aHeader, aJitter);
 }
-
-void WebrtcAudioConduit::OnRtcpBye() {
-  RefPtr<WebrtcAudioConduit> self = this;
-  NS_DispatchToMainThread(media::NewRunnableFrom([self]() mutable {
-    MOZ_ASSERT(NS_IsMainThread());
-    if (self->mRtcpEventObserver) {
-      self->mRtcpEventObserver->OnRtcpBye();
-    }
-    return NS_OK;
-  }));
-}
-
-void WebrtcAudioConduit::OnRtcpTimeout() {
-  RefPtr<WebrtcAudioConduit> self = this;
-  NS_DispatchToMainThread(media::NewRunnableFrom([self]() mutable {
-    MOZ_ASSERT(NS_IsMainThread());
-    if (self->mRtcpEventObserver) {
-      self->mRtcpEventObserver->OnRtcpTimeout();
-    }
-    return NS_OK;
-  }));
-}
+*/
 
 void WebrtcAudioConduit::SetRtcpEventObserver(
     mozilla::RtcpEventObserver* observer) {
@@ -364,12 +364,12 @@
 MediaConduitErrorCode WebrtcAudioConduit::Init() {
   CSFLogDebug(LOGTAG, "%s this=%p", __FUNCTION__, this);
   MOZ_ASSERT(NS_IsMainThread());
-
-  if (!(mPtrVoEBase = webrtc::VoEBase::GetInterface(GetVoiceEngine()))) {
-    CSFLogError(LOGTAG, "%s Unable to initialize VoEBase", __FUNCTION__);
-    return kMediaConduitSessionNotInited;
-  }
-
+  /*
+    if (!(mPtrVoEBase = webrtc::VoEBase::GetInterface(GetVoiceEngine()))) {
+      CSFLogError(LOGTAG, "%s Unable to initialize VoEBase", __FUNCTION__);
+      return kMediaConduitSessionNotInited;
+    }
+  */
   CreateChannels();
 
   CSFLogDebug(LOGTAG, "%s AudioSessionConduit Initialization Done (%p)",
@@ -495,8 +495,9 @@
     webrtc::SdpAudioFormat format(codec->mName, codec->mFreq, codec->mChannels,
                                   parameters);
     mRecvStreamConfig.decoder_map.emplace(codec->mType, format);
-
-    mRecvStreamConfig.voe_channel_id = mRecvChannel;
+    /*
+        mRecvStreamConfig.voe_channel_id = mRecvChannel;
+    */
     success = true;
   }  // end for
 
@@ -556,7 +557,7 @@
     }
 
     // MID RTP header extension
-    if (extension.uri == webrtc::RtpExtension::kMIdUri) {
+    if (extension.uri == webrtc::RtpExtension::kMidUri) {
       if (!isSend) {
         // TODO(bug 1405495): Why do we error out for csrc-audio-level, but not
         // mid?
@@ -575,19 +576,17 @@
   }
 
   currentExtensions = filteredExtensions;
-
-  if (isSend) {
-    mSendChannelProxy->SetSendAudioLevelIndicationStatus(ssrcAudioLevelId != -1,
-                                                         ssrcAudioLevelId);
-    mSendChannelProxy->SetSendMIDStatus(midId != -1, midId);
-  } else {
-    mRecvChannelProxy->SetReceiveAudioLevelIndicationStatus(
-        ssrcAudioLevelId != -1, ssrcAudioLevelId);
-    mRecvChannelProxy->SetReceiveCsrcAudioLevelIndicationStatus(
-        csrcAudioLevelId != -1, csrcAudioLevelId);
-    // TODO(bug 1405495): recv mid support
-  }
-
+  /*
+    if (isSend) {
+      mSendChannelProxy->SetSendAudioLevelIndicationStatus(ssrcAudioLevelId !=
+    -1, ssrcAudioLevelId); mSendChannelProxy->SetSendMIDStatus(midId != -1,
+    midId); } else { mRecvChannelProxy->SetReceiveAudioLevelIndicationStatus(
+          ssrcAudioLevelId != -1, ssrcAudioLevelId);
+      mRecvChannelProxy->SetReceiveCsrcAudioLevelIndicationStatus(
+          csrcAudioLevelId != -1, csrcAudioLevelId);
+      // TODO(bug 1405495): recv mid support
+    }
+  */
   if (isSend) {
     RecreateSendStreamIfExists();
   } else {
@@ -633,10 +632,12 @@
   }
 
   // Insert the samples
-  mPtrVoEBase->audio_transport()->PushCaptureData(
-      mSendChannel, audio_data,
-      sizeof(audio_data[0]) * 8,  // bits
-      samplingFreqHz, channels, lengthSamples);
+  /*
+    mPtrVoEBase->audio_transport()->PushCaptureData(
+        mSendChannel, audio_data,
+        sizeof(audio_data[0]) * 8,  // bits
+        samplingFreqHz, channels, lengthSamples);
+  */
   // we should be good here
   return kMediaConduitNoError;
 }
@@ -678,8 +679,9 @@
 
   size_t lengthSamplesAllowed = lengthSamples;
   lengthSamples = 0;  // output paramter
-
-  mRecvChannelProxy->GetAudioFrameWithInfo(samplingFreqHz, &mAudioFrame);
+                      /*
+                        mRecvChannelProxy->GetAudioFrameWithInfo(samplingFreqHz, &mAudioFrame);
+                      */
   numChannels = mAudioFrame.num_channels_;
 
   if (numChannels == 0) {
@@ -1040,7 +1042,9 @@
     mSendStream = nullptr;
   }
   // Destroying the stream unregisters the transport
-  mSendChannelProxy->RegisterTransport(nullptr);
+  /*
+    mSendChannelProxy->RegisterTransport(nullptr);
+  */
 }
 
 MediaConduitErrorCode WebrtcAudioConduit::CreateSendStream() {
@@ -1065,7 +1069,9 @@
     mRecvStream = nullptr;
   }
   // Destroying the stream unregisters the transport
-  mRecvChannelProxy->RegisterTransport(nullptr);
+  /*
+    mRecvChannelProxy->RegisterTransport(nullptr);
+  */
 }
 
 MediaConduitErrorCode WebrtcAudioConduit::CreateRecvStream() {
@@ -1122,8 +1128,8 @@
   // Bug 1499796 - we need to get passed the time the packet was received
   webrtc::PacketReceiver::DeliveryStatus status =
       mCall->Call()->Receiver()->DeliverPacket(
-          webrtc::MediaType::AUDIO, static_cast<const uint8_t*>(data), len,
-          webrtc::PacketTime());
+          webrtc::MediaType::AUDIO,
+          rtc::CopyOnWriteBuffer(static_cast<const uint8_t*>(data), len), -1);
 
   if (status != webrtc::PacketReceiver::DELIVERY_OK) {
     CSFLogError(LOGTAG, "%s DeliverPacket Failed, %d", __FUNCTION__, status);
@@ -1135,41 +1141,41 @@
 
 MediaConduitErrorCode WebrtcAudioConduit::CreateChannels() {
   MOZ_ASSERT(NS_IsMainThread());
+  /*
+    if ((mRecvChannel = mPtrVoEBase->CreateChannel()) == -1) {
+      CSFLogError(LOGTAG, "%s VoiceEngine Channel creation failed",
+    __FUNCTION__); return kMediaConduitChannelError;
+    }
+    mRecvStreamConfig.voe_channel_id = mRecvChannel;
 
-  if ((mRecvChannel = mPtrVoEBase->CreateChannel()) == -1) {
-    CSFLogError(LOGTAG, "%s VoiceEngine Channel creation failed", __FUNCTION__);
-    return kMediaConduitChannelError;
-  }
-  mRecvStreamConfig.voe_channel_id = mRecvChannel;
-
-  if ((mSendChannel = mPtrVoEBase->CreateChannel()) == -1) {
-    CSFLogError(LOGTAG, "%s VoiceEngine Channel creation failed", __FUNCTION__);
-    return kMediaConduitChannelError;
-  }
-  mSendStreamConfig.voe_channel_id = mSendChannel;
+    if ((mSendChannel = mPtrVoEBase->CreateChannel()) == -1) {
+      CSFLogError(LOGTAG, "%s VoiceEngine Channel creation failed",
+    __FUNCTION__); return kMediaConduitChannelError;
+    }
+    mSendStreamConfig.voe_channel_id = mSendChannel;
 
-  webrtc::VoiceEngineImpl* vei;
-  vei = static_cast<webrtc::VoiceEngineImpl*>(GetVoiceEngine());
-  mRecvChannelProxy = vei->GetChannelProxy(mRecvChannel);
-  if (!mRecvChannelProxy) {
-    CSFLogError(LOGTAG, "%s VoiceEngine Send ChannelProxy creation failed",
-                __FUNCTION__);
-    return kMediaConduitChannelError;
-  }
+    webrtc::VoiceEngineImpl* vei;
+    vei = static_cast<webrtc::VoiceEngineImpl*>(GetVoiceEngine());
+    mRecvChannelProxy = vei->GetChannelProxy(mRecvChannel);
+    if (!mRecvChannelProxy) {
+      CSFLogError(LOGTAG, "%s VoiceEngine Send ChannelProxy creation failed",
+                  __FUNCTION__);
+      return kMediaConduitChannelError;
+    }
 
-  mRecvChannelProxy->SetRtpPacketObserver(this);
-  mRecvChannelProxy->SetRtcpEventObserver(this);
-  mRecvChannelProxy->RegisterTransport(this);
+    mRecvChannelProxy->SetRtpPacketObserver(this);
+    mRecvChannelProxy->SetRtcpEventObserver(this);
+    mRecvChannelProxy->RegisterTransport(this);
 
-  mSendChannelProxy = vei->GetChannelProxy(mSendChannel);
-  if (!mSendChannelProxy) {
-    CSFLogError(LOGTAG, "%s VoiceEngine ChannelProxy creation failed",
-                __FUNCTION__);
-    return kMediaConduitChannelError;
-  }
-  mSendChannelProxy->SetRtpPacketObserver(this);
-  mSendChannelProxy->RegisterTransport(this);
-
+    mSendChannelProxy = vei->GetChannelProxy(mSendChannel);
+    if (!mSendChannelProxy) {
+      CSFLogError(LOGTAG, "%s VoiceEngine ChannelProxy creation failed",
+                  __FUNCTION__);
+      return kMediaConduitChannelError;
+    }
+    mSendChannelProxy->SetRtpPacketObserver(this);
+    mSendChannelProxy->RegisterTransport(this);
+  */
   return kMediaConduitNoError;
 }
 
@@ -1178,15 +1184,18 @@
   mMutex.AssertCurrentThreadOwns();
 
   if (mSendChannel != -1) {
-    mSendChannelProxy = nullptr;
-    mPtrVoEBase->DeleteChannel(mSendChannel);
+    /*
+        mSendChannelProxy = nullptr;
+        mPtrVoEBase->DeleteChannel(mSendChannel);
+    */
     mSendChannel = -1;
   }
 
   if (mRecvChannel != -1) {
-    mRecvChannelProxy->SetRtcpEventObserver(nullptr);
-    mRecvChannelProxy = nullptr;
-    mPtrVoEBase->DeleteChannel(mRecvChannel);
+    /*
+        mRecvChannelProxy = nullptr;
+        mPtrVoEBase->DeleteChannel(mRecvChannel);
+    */
     mRecvChannel = -1;
   }
 }